plutus-wallet-api/ledger/Ledger/Typed/TypeUtils.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE Rank2Types            #-}
+module Ledger.Typed.TypeUtils where
+
+import           Data.Kind
+import           Data.Proxy
+
+-- | A heterogeneous list where every element is wrapped with the given functor.
+data HListF (f :: Type -> Type) (l :: [Type]) where
+    HNilF  :: HListF f '[]
+    HConsF :: f e -> HListF f l -> HListF f (e ': l)
+
+-- | Turn a 'HListF' into a homogeneous list. Requires a very polymorphic function, likely something like 'id' or 'coerce'.
+hfOut :: forall o f (ts :: [Type]) . (forall a . f a -> o) -> HListF f ts -> [o]
```

Comment:
`id` won't work, because you need a function that "forgets" the `a` type parameter and `id` never forgets.